### PR TITLE
Prevent assert if stack trace has function pointer

### LINKED ISSUE #63785: NativeAOT: Application crash during exception stack building
I receive following exception.

```
 native\ICustomMarshaler_TargetWindows
Unhandled Exception: EETypeRva:0x0074DFF0(System.Diagnostics.DebugProvider.DebugAssertException):    at System.Diagnostics.DebugProvider.Fail(String, String) + 0x4b
   at System.Diagnostics.Debug.Fail(String, String) + 0x45
   at System.Diagnostics.Debug.Assert(Boolean, String, String) + 0x3c
   at System.Diagnostics.Debug.Assert(Boolean) + 0x27
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitTypeName(Handle, Boolean) + 0x4f9
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitTypeSpecificationName(TypeSpecificationHandle, Boolean) + 0x6f
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitTypeName(Handle, Boolean) + 0x16f
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitTypeVector(HandleCollection) + 0xb7
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitMethodParameters(MethodSignature) + 0x78
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitMethodReferenceName(MemberReferenceHandle) + 0xe4
   at Internal.StackTraceMetadata.MethodNameFormatter.EmitMethodName(Handle) + 0x8d
   at Internal.StackTraceMetadata.MethodNameFormatter.FormatMethodName(MetadataReader, Handle) + 0xa2
   at Internal.StackTraceMetadata.StackTraceMetadata.PerModuleMethodNameResolver.GetMethodNameFromRvaIfAvailable(Int32) + 0xa2
   at Internal.StackTraceMetadata.StackTraceMetadata.GetMethodNameFromStartAddressIfAvailable(IntPtr) + 0x189
   at Internal.StackTraceMetadata.StackTraceMetadata.StackTraceMetadataCallbacksImpl.TryGetMethodNameFromStartAddress(IntPtr) + 0x26
   at Internal.DeveloperExperience.DeveloperExperience.CreateStackTraceString(IntPtr, Boolean) + 0x121
   at System.Diagnostics.StackFrame.AppendToStackTrace(StringBuilder) + 0x124
   at System.Diagnostics.StackTrace.ToString(StackTrace.TraceFormat, StringBuilder) + 0x96
   at System.Diagnostics.StackTrace.ToString(StackTrace.TraceFormat) + 0x5a
   at System.Exception.GetStackTrace() + 0x4b
   at System.Exception.get_StackTrace() + 0xae
   at System.Exception.ToString() + 0x14e
   at System.Runtime.CompilerServices.DefaultInterpolatedStringHandler.AppendFormatted[T](T) + 0x235
   at System.Runtime.InteropServices.Tests.ICustomMarshalerTests.Main(String[]) + 0x12d
   at System.Runtime.EH.DispatchEx(StackFrameIterator&, EH.ExInfo&, UInt32) + 0x576
   at System.Runtime.EH.RhThrowEx(Object, EH.ExInfo&) + 0x8d
   at ICustomMarshaler_TargetWindows!<BaseAddress>+0x51c607
   at ICustomMarshaler_TargetWindows!<BaseAddress>+0x51c6a0
```

Exception happens when `typeHandle.HandleType` from this line has value `HandleType.FunctionPointerSignature`
https://github.com/dotnet/runtime/blob/b2dc37ba181a7fa4427e717eab819ba3543d0ae4/src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/MethodNameFormatter.cs#L212

so I assume that if stack trace has function pointer signature then whole app crash. 
I will work on the fix. Have just one question, should I wrote test for that, and where if any.

### PR DESCRIPTION
**Pull request by kant2002**
Closes #63785

### CODE REVIEW

File: src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/MethodNameFormatter.cs [hunk 1]
```diff
@@ -255,8 +255,12 @@ private void EmitTypeName(Handle typeHandle, bool namespaceQualified)
                     EmitString(typeHandle.ToGenericParameterHandle(_metadataReader).GetGenericParameter(_metadataReader).Name);
                     break;
 
+                case HandleType.FunctionPointerSignature:
+                    EmitFunctionPointerTypeName(typeHandle.ToFunctionPointerSignatureHandle(_metadataReader));
+                    break;
+
                 default:
-                    Debug.Assert(false);
+                    Debug.Assert(false, $"Type handle {typeHandle.HandleType} does not handled");
```
Comment (Wraith2) [inline src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/MethodNameFormatter.cs:263]: "is" or "was" not handled seems more appropriate.

File: src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/MethodNameFormatter.cs [hunk 2]
```diff
@@ -407,6 +411,21 @@ private void EmitPointerTypeName(PointerSignatureHandle pointerSigHandle)
...
         }
 
+        /// <summary>
+        /// Emit function pointer type.
+        /// </summary>
+        /// <param name="functionPointerSigHandle">Function pointer type specification signature handle</param>
+        private void EmitFunctionPointerTypeName(FunctionPointerSignatureHandle functionPointerSigHandle)
+        {
+            FunctionPointerSignature functionPointerSig = _metadataReader.GetFunctionPointerSignature(functionPointerSigHandle);
+            _outputBuilder.Append("delegate*<");
```
Comment (MichalStrehovsky) [inline src/coreclr/nativeaot/System.Private.StackTraceMetadata/src/Internal/StackTraceMetadata/MethodNameFormatter.cs:421]: This should emit whatever CoreCLR will emit for this, which is probably not delegate*. Not at my PC anymore but I'm guessing IntPtr. Thanks for looking into this!

**Review by MichalStrehovsky — COMMENTED** [review]
Thank you! Could you add a quick test to src\libraries\System.Diagnostics.StackTrace\tests\StackTraceTests.cs? Something that just checks that `void Test(delegate*<void> unused) { new StringReader(new StackTrace().ToString()).ReadLine().Contains("IntPtr") == true }`. This behavior is likely to change and I'm pretty sure we would forget to update this logic when that happens.

File: src/libraries/System.Diagnostics.StackTrace/tests/StackTraceTests.cs [hunk 3]
```diff
@@ -312,6 +312,14 @@ public void ToString_NullFrame_ThrowsNullReferenceException()
             Assert.Equal(Environment.NewLine, stackTrace.ToString());
         }
 
+        [Fact]
+        public unsafe void ToString_FunctionPointerSignature()
+        {
+            // This is sepate from ToString_Invoke_ReturnsExpected since unsafe cannot be used for iterators
+            var stackTrace = FunctionPointerParameter(null);
```
Comment (deeprobin) [inline src/libraries/System.Diagnostics.StackTrace/tests/StackTraceTests.cs:319]: Usage of explicit type (for reference: [Coding Guidelines](https://github.com/dotnet/runtime/blob/25d999b5a6711fc6d2ffc42fa38a58ec865a7232/docs/coding-guidelines/coding-style.md?plain=1#L26)))

```suggestion
            StackTrace stackTrace = FunctionPointerParameter(null);
```

**Review by deeprobin — APPROVED** [review]
LGTM. Have only a few small style changes proposed.

File: src/libraries/System.Diagnostics.StackTrace/tests/StackTraceTests.cs (hunk 3 above)
Comment (kant2002) [reply src/libraries/System.Diagnostics.StackTrace/tests/StackTraceTests.cs:319]: I see `var` style used a lot in this file. I would say most of the time.

File: src/libraries/System.Diagnostics.StackTrace/tests/StackTraceTests.cs [hunk 4]
```diff
@@ -312,6 +312,21 @@ public void ToString_NullFrame_ThrowsNullReferenceException()
...
+            if (PlatformDetection.IsMonoRuntime)
+            {
+                Assert.Contains("System.Diagnostics.Tests.StackTraceTests.FunctionPointerParameter(MonoFNPtrFakeClass x)", stackTrace.ToString());
```
Comment (MichalStrehovsky) [inline src/libraries/System.Diagnostics.StackTrace/tests/StackTraceTests.cs:322]: The Mono behavior is a bug that we'll want to fix eventually. Could you instead add `[ActiveIssue("https://github.com/dotnet/runtime/issues/11354", TestRuntimes.Mono)]` to this test?

(It doesn't make sense to cut a new bug for it; we have plenty of Mono bugs - Mono will either be fixed when 11354 is fixed, or there will be a new Mono-specific bug that we'll need to open as part of fixing 11354 to block all the failing tests added at that point.)

**Review by MichalStrehovsky — APPROVED** [review]
Thank you!

**Pull request merged**